java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerSessionMiddleware.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.flight;
+
+/**
+ * Middleware to add an optional set-session/session header.
+ */
+public class ServerSessionMiddleware implements FlightServerMiddleware {
+
+  /**
+   * Factory class to provide instances of ServerSessionMiddleware for each call.
+   */
+  public static class Factory implements FlightServerMiddleware.Factory<ServerSessionMiddleware> {
+    private final ServerSessionHandler sessionHandler;
+
+    public Factory(ServerSessionHandler sessionHandler) {
+      this.sessionHandler = sessionHandler;
+    }
+
+    @Override
+    public ServerSessionMiddleware onCallStarted(CallInfo info, CallHeaders incomingHeaders,
+                                                 RequestContext context) {
+
+      String session = sessionHandler.getSession(incomingHeaders);
```

Comment:
I feel getSession should take in the session token instead of all headers, since we explicitly have this middleware send the token.